pkg/odo/cli/component/test.go

```diff
@@ -0,0 +1,97 @@
+package component
+
+import (
+	"fmt"
+	"path/filepath"
+	"reflect"
+
+	"github.com/pkg/errors"
+
+	devfileParser "github.com/openshift/odo/pkg/devfile/parser"
+	projectCmd "github.com/openshift/odo/pkg/odo/cli/project"
+	"github.com/openshift/odo/pkg/odo/genericclioptions"
+	odoutil "github.com/openshift/odo/pkg/odo/util"
+	"github.com/openshift/odo/pkg/odo/util/completion"
+	"github.com/spf13/cobra"
+	"k8s.io/kubectl/pkg/util/templates"
+)
+
+// RecommendedCommandName is the recommended command name
+const TestRecommendedCommandName = "test"
+
+// TestOptions encapsulates the options for the odo command
+type TestOptions struct {
+	commandName      string
+	componentContext string
+	devfilePath      string
+	show             bool
+	devObj           devfileParser.DevfileObj
+	*genericclioptions.Context
+}
+
+var testExample = templates.Examples(`
+  # Run default test command
+  %[1]s
+
+  # Run a specific test command
+  %[1]s --test-command <command name>
+
+`)
+
+// NewTestOptions creates a new TestOptions instance
+func NewTestOptions() *TestOptions {
+	return &TestOptions{}
+}
+
+// Complete completes TestOptions after they've been created
+func (to *TestOptions) Complete(name string, cmd *cobra.Command, args []string) (err error) {
+	to.devfilePath = filepath.Join(to.componentContext, DevfilePath)
+	to.Context = genericclioptions.NewDevfileContext(cmd)
+
+	return
+}
+
+// Validate validates the TestOptions based on completed values
+func (to *TestOptions) Validate() (err error) {
+	devObj, err := devfileParser.ParseAndValidate(to.devfilePath)
+	if err != nil {
+		return errors.Wrap(err, "fail to parse devfile")
+	}
+	if reflect.DeepEqual(devObj.Ctx.GetApiVersion(), "1.0.0") {
+		return fmt.Errorf("'odo test' is not supported in devfile 1.0.0")
+	}
+	to.devObj = devObj
+	return
+}
+
+// Run contains the logic for the odo command
+func (to *TestOptions) Run() (err error) {
+	return to.RunTestCommand()
+}
+
+// NewCmdTest implements the odo test command
+func NewCmdTest(name, fullName string) *cobra.Command {
+	to := NewTestOptions()
+	testCmd := &cobra.Command{
+		Use:     name,
+		Short:   "Run the test command defined in devfile",
+		Long:    "Run the test command defined in devfile",
+		Example: fmt.Sprintf(testExample, fullName),
+		Args:    cobra.MaximumNArgs(1),
```

Comment:
Why are we accepting a argument? It doesn't seem to anything with it in the code.
```
[mrinaldas@localhost nodejs-ex]$ odo test garbage
I0714 10:58:53.346770   16885 cert_rotation.go:137] Starting client certificate rotation controller

Executing devfile test command for component nodejs
 ✓  Executing test command "npm test" [265ms]
[mrinaldas@localhost nodejs-ex]$ odo test blhaaddaskjha
I0714 10:58:56.330496   16908 cert_rotation.go:137] Starting client certificate rotation controller

Executing devfile test command for component nodejs
 ✓  Executing test command "npm test" [269ms]
```

Reply:
The argument is the specific command name we want to run if we don't want to run the default one. 
And if the command does not exist, we run the default command.

Reply:
You are right. This should be 0 args

Reply:
updated